### PR TITLE
test(serialization): cover multidimensional array copying

### PR DESCRIPTION
**Pull request by ReubenBond**
## Problem

`MultiDimensionalArrayCopier<T>.DeepCopy` had no coverage despite complexity 22 and a reported CRAP score of 506. Its lower-bound handling also assumed zero-based arrays, and shallow-copyable arrays were not recorded in the copy context.

## Solution

Add focused multidimensional array tests covering ranks 1, 2, and 3+, shallow and reference elements, repeated aliases, cycles, empty dimensions, extreme CLR bounds, round-trip independence, supported-type boundaries, and the non-zero-lower-bound serialization failure contract.

Preserve lower bounds during deep copy, record shallow array copies so aliases remain stable, use offset-based loops which remain correct at `int.MinValue`/`int.MaxValue`, and reject non-zero lower bounds during serialization explicitly because the wire format records lengths only.

The targeted coverage run raises `MultiDimensionalArrayCopier<T>` and `DeepCopy` from 0% line / 0% branch coverage to 100% line / 100% branch coverage, reducing the method CRAP score from 506 to 22.

This is a mergeable first slice of #10859. The repository-wide build currently encounters the independent main-branch failure tracked by #10856; #10878 contains that fix.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/orleans/pull/10880)

### CODE REVIEW

File: test/Orleans.Serialization.UnitTests/MultiDimensionalArrayTests.cs [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+        var rankTwoResult = Copy(rankTwoEmpty);
+
+        Assert.Equal(int.MaxValue, rankOneResult.GetLowerBound(0));
+        Assert.Equal(int.MaxValue, rankOneResult.GetUpperBound(0));
+        Assert.Equal(65, Assert.IsType<MyValue>(rankOneResult.GetValue(int.MaxValue)).Value);
+        Assert.Equal(int.MaxValue, rankTwoResult.GetLowerBound(0));
+        Assert.Equal(int.MaxValue, rankTwoResult.GetUpperBound(0));
+        Assert.Equal(int.MinValue, rankTwoResult.GetLowerBound(1));
+        Assert.Empty(rankTwoResult);
+    }
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply test/Orleans.Serialization.UnitTests/MultiDimensionalArrayTests.cs:158]: Fixed in `0f11bcd746`. The rank-two case now uses a length-one second dimension at `int.MinValue`, writes a value at the extreme indices, and asserts both preserved bounds and the copied value, avoiding zero-length upper-bound underflow differences across runtimes.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: src/Orleans.Serialization/Codecs/MultiDimensionalArrayCodec.cs [hunk 2]
```diff
@@ -36,6 +36,11 @@ public MultiDimensionalArrayCodec(IFieldCodec<int[]> intArrayCodec, IFieldCodec<
         /// <inheritdoc/>
         public void WriteField<TBufferWriter>(ref Writer<TBufferWriter> writer, uint fieldIdDelta, [System.Diagnostics.CodeAnalysis.AllowNull] Type expectedType, [System.Diagnostics.CodeAnalysis.AllowNull] object? value) where TBufferWriter : IBufferWriter<byte>
         {
+            if (value is Array input)
+            {
+                EnsureZeroLowerBounds(input);
+            }
+
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Serialization/Codecs/MultiDimensionalArrayCodec.cs]: Fixed in `0f11bcd746`. `TryWriteReferenceField` now runs first, and lower-bound validation occurs only for the first full array write, preserving the failure contract while avoiding repeated scans for reference fields.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).
